lib/FunctionIsolation/EnforceABI.cpp

```diff
@@ -443,23 +492,52 @@ void EnforceABIImpl::handleRoot() {
     // Collect arguments
     IRBuilder<> Builder(Invoke);
     std::vector<Value *> Arguments;
+    std::vector<GlobalVariable *> ReturnCSVs;
     for (auto &P : Function.RegisterSlots) {
       GlobalVariable *CSV = P.first;
+
       if (shouldEmit(P.second.Argument))
         Arguments.push_back(Builder.CreateLoad(CSV));
+
+      if (shouldEmit(P.second.ReturnValue))
+        ReturnCSVs.push_back(CSV);
+    }
+
+    BasicBlock *NormalDest;
+    if (!ReturnCSVs.empty()) {
+      NormalDest = BasicBlock::Create(Context,
+                                      BB.getName() + "__extract_retval",
+                                      Root,
+                                      &BB);
+      NormalDest->moveAfter(&BB);
+    } else {
+      NormalDest = Invoke->getNormalDest();
     }
 
     // Create the new invoke with the appropriate arguments
-    auto *NewInvoke = Builder.CreateInvoke(Callee,
-                                           Invoke->getNormalDest(),
-                                           Invoke->getUnwindDest(),
-                                           Arguments);
-    NewInvoke->setIsNoInline();
+    auto *Result = Builder.CreateInvoke(Callee,
+                                        NormalDest,
+                                        Invoke->getUnwindDest(),
+                                        Arguments);
+    Result->setIsNoInline();
```

Comment:
Is this necessary?

Reply:
Yes, I believe we always want the call-site in the entry basic blocks in `root`.

Reply:
Why? Does it ever happen on any of the pipelines we have that it gets inlined?
If we launch `-O2` to optimized runtime performance I don't see why we should constrain the optimizer in this way.

Reply:
I don't see why we should prevent any optimizations there. Note that `setIsNoInline` was already there before such changes and I thought there were other reasons why it was initially introduced.

Reply:
Uhm, I thought I checked but you're right, it was there before.
There's probably a reason for that then, let's leave it there. Maybe we can ask @andrealinux1.
Anyway, `Invoke` seems like a better name than `Result`, restore it or change it to something more meaningful.

Reply:
Changed it to `NewInvoke` since we are replacing the old one.

Reply:
Ok, so the journey is a bit convoluted here. I think you (@antoniofrighetto) borrowed these lines creating the invokes from the corresponding ones in the `FunctionIsolation`, specifically where we create the first `invoke`s that are in charge of switching the execution from the non isolated realm to the isolated one, and that constitute the substituted body of each function entry block.

There, we added the `noinline` (commit af6f963d7765bee9aeb3502d63896187e0fdf316), in order to avoid the `prune-eh` optimization pass to remove these `invoke`s and trasforming them into `call`s. From the commit message:
```
Improve exception handling
    
    When optimizing a module to which function isolation has been applied the
    `prune-eh` pass lead to several issues, specifically:
    
    * `support.c` is now compiled with support for exceptions, to prevent
      `raise_exception_helper` from being marked `nounwind` during the
      link phase.
    * Added a fake `ret` at the end of the `catchblock` to avoid promotion
      of `invoke`s to regular `call`s.
    * Marked the `invoke` instructions as `noinline`.
``` 

Now, reading the `prune-eh` pass, it seems to me that maybe that was due to the fact that, before @antoniofrighetto really understood how the personality function works (and fixed it in 76a6403ebdf2acb7fb289159725c0917f5e12f40),  we didn't really use the unwind block (I don't know if you remember the story about the `ExceptionFlag`), and therefore the `prune-eh` interprocedural analysis decided to promote the `invoke`s to `calls`.

I expect that now, the `prune-eh` is doing things correctly, so it may be worth a try to see if the `invoke`s are still disappearing after that change.

Reply:
This last analysis is correct, however, note that the `invoke` creation as well as the `noinline` attributes were already added by @aleclearmind. Our contribution there was simply concerned with extracting the return values out of the call in case we would be returning from an isolated function to `root` (e.g., after an exception has been managed). Now after installing the personality routine and with the removal of `ExceptionFlag`, I don't see why the `noinline` should be kept.

Reply:
@andrealinux1 saves the day!
Can you give it a shot in a working example you have @antoniofrighetto?

Reply:
OK, please test on a working example and make a new commit to drop this.

Keep in mind that PRs and code review are an opportunity to improve the existing code base.
So even if the bug was already there, it's worth addressing it since we're already working on this.

Reply:
(thanks @andrealinux1!)

Reply:
Fully agree. Btw, I do remember those lines driving us crazy back then – while debugging an enforce'd binary –, so yeah, checking it again surely doesn't harm.